Adds multi promise sources

esirK opened the pull request:
## Description

- Allows multiple sources from which promises will be fetched.
- Adds JSON file as a source of promises.
- 
Fixes # (issue)

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Desktop Screenshots
<img width="1430" alt="Screenshot 2021-07-30 at 12 04 23" src="https://user-images.githubusercontent.com/13068580/127629605-832c51c6-eef3-4488-bf62-66f55180eec6.png">

## Mobile Screenshots

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

kilemensi reviewed (commented):
👍🏾 

---------------

I think there are multiple ways we can handle JSON file source:

1. `public` folder: Since the contents are public anyway, we can store the file in public with other assets. This would mean updating promises implies redeploying.
2. JSON file passed via a URL environmental variable. This means a file could live anywhere else outside the site and we only need to load it once (any may be store a cache copy in public folder so we don't have to fetch it every time).

For file-based source (gSheet included), the file source not only needs to include promises, but everything needed for /promises page e.g. categories, statuses, etc. So for JSON this would be an object-key style, for gSheet this could be multiple worksheets each with one purpose, etc.

We need a good interface that will offer basic access regardless of the source e.g. how to fetch statuses, how to fetch key promises, etc.

kilemensi commented on src/pages/promises/index.js:
```diff
@@ -112,12 +113,20 @@ export async function getStaticProps({ locale }) {
+    }),
+  };
+
+  const sourceLib = process.env.SOURCE_LIB || "Check";
+
+  const api = promiseSourceLibs[sourceLib];
```
Rather than doing this on every page, can't the  `promisesSourceLibs` (probably renamed to `promisesApi`) check the `env` on initialisation? i.e. one only need to call `promisesApi()`.

esirK replied:
@kilemensi  as I continue working on this, I think the functionality is ready for re-review.

kilemensi replied:
☝🏾 still not done @esirK

esirK replied:
this has now been done.

kilemensi commented on src/lib/json_source/json_source_client.js:
```diff
@@ -0,0 +1,63 @@
+          });
+        case "GET_KEY_PROMISES":
+          return new Promise((resolve, reject) => {
+            if (promises.promises === undefined) {
+              reject(new Error("No promises defined"));
+            }
+            const keyPromises = promises.promises.filter(
+              (p) => `${p.key_promise}` === "True"
+            );
```
```suggestion
            const keyPromises = promises?.promises?.filter(
              (p) => `${p.key_promise}` === "True"
            ) ?? [];
```

I think it's best we return empty promises whenever we can rather than rejecting Promise or throwing errors.

PS: Any reason why we're defining Promises when we already have the JSON file imported?

esirK replied:
@kilemensi  for 

>PS: Any reason why we're defining Promises when we already have the JSON file imported?

I'm not sure I fully understand your inquiry. Do you mean doing just 

`import  "../../../public/promises.json";`
  
should be enough?

kilemensi replied:
`Promise` as in `new Promise` i.e. going async when we already have everything loaded in memory. We're not doing any network calls, etc. so I don't see Promises adding anything here.

esirK replied:
Oooh. I see

esirK replied:
I think it will be useful once we migrate to using URLs for the JSON file source,

esirK replied:
@kilemensi  Given that the final goal is to load JSON from a server, I think we can maintain this async functionality. What do you think?

kilemensi replied:
Not sure @esirK ... Are we going to fetch the JSON file periodically from the URL and store it in the public folder for catching? This means all our processing will be done from local JSON file which can load in memory.

esirK replied:
Yeah. In that case, we really don't need the async.

kilemensi reviewed (commented):
👍🏾 

----

1. Each promise status should have its own colour. They all seem to have the same colour 
![Screenshot from 2021-08-11 12-04-03](https://user-images.githubusercontent.com/1779590/129001967-19f0b257-4a33-4ccf-8a6c-1444895ee3cc.png)
2. Key promises don't show image. Are we getting the promise events as well?
![Screenshot from 2021-08-11 12-04-56](https://user-images.githubusercontent.com/1779590/129002066-9dd2fbe1-f20d-4af9-ac21-6ae80c012b99.png)
3. We have 3 promises but no computation on the hero.
![Screenshot from 2021-08-11 12-05-22](https://user-images.githubusercontent.com/1779590/129002153-72436b86-5735-456b-93cb-cee36dbbcb18.png)

esirK commented:
> 1. Each promise status should have its own color. They all seem to have the same color

yes. I'll fix this up. I think I had added some and removed them at some point. 

Could you share with me the latest promise tracker designs so I can compare all these details?
 
> 2. Key promises don't show image. Are we getting the promise events as well?

Currently no. with the designs I'll be able to update this.

kilemensi commented:
Designs are on [Trello](https://trello.com/c/XzLq5arR/23-link-to-prototype) @esirK 

Check API https://check-api.checkmedia.org/graphiql

kilemensi commented on public/promises.json:
```diff
@@ -0,0 +1,78 @@
+      "href": "/promises/1/health",
+      "id": 1,
+      "key_promise": "True",
+      "relatedPromises": [
+        {
+          "id": 2,
+          "tags": ["health", "war"],
+          "title": "Build School",
+          "slug": "Build-School",
+          "href": "/promises/2/Build-School",
+          "date": "29/07/2021",
+          "status": { "title": "Complete" },
+          "image": "https://upload.wikimedia.org/wikipedia/commons/d/dd/Uhuru_Kenyatta.jpg"
+        }
+      ],
```
relatedPromises will have to "computed" based on tags/categories/etc.

esirK replied:
@kilemensi concerning this, 
if 2 promises have the same tags/categories then they should be treated as related? Or are there minimum similarities that should we use?

kilemensi replied:
Nah, I think we should keep it as simple as possible for now... but yes, the more tags/categories they share the better/more related @esirK

kilemensi commented on src/lib/json_source/json_source_client.js:
```diff
@@ -0,0 +1,47 @@
+        case "GET_PROMISE":
+          try {
+            const promise = promises.promises.filter(
+              (p) => `${p.id}` === id
```
Lets match using [title slugs](https://github.com/CodeForAfrica/PromiseTracker/blob/476c2bbe7c46695c39163a02499f8bdd31da7645/src/utils.js#L1)

kilemensi commented on src/lib/json_source/json_source_client.js:
```diff
@@ -0,0 +1,47 @@
+          }
+        case "GET_KEY_PROMISES": {
+          const keyPromises =
+            promises?.promises?.filter((p) => `${p.key_promise}` === "True") ??
```
Lets use camelCase and `keyPromise` should just be a boolean

kilemensi commented on src/lib/json_source/json_source_client.js:
```diff
@@ -0,0 +1,47 @@
+        }
+        case "GET_PROMISES_BY_CATEGORY": {
+          const promisesByCat = promises.promises.filter(
+            (p) => p.category === category
```
comparison should handle casing

kilemensi reviewed (commented):
👍🏾 

----

Other than naming, I think it's looking good.

kilemensi commented on src/lib/jsonSource/jsonSourceClient.js:
```diff
@@ -0,0 +1,60 @@
+import { slugify } from "@/promisetracker/utils";
+
+const JsonSourceClient = () => {
+  // Sulugify all promises
```
```suggestion
  // Slugify all promises' titles
```

kilemensi commented on src/lib/jsonSource/jsonSourceClient.js:
```diff
@@ -0,0 +1,60 @@
+          return allPromises.slice(0, limit);
+        case "GET_PROMISE":
+          try {
+            const promise = allPromises.filter((p) => `${p.slug}` === slug)[0];
```
Lets find by `id`... titles can be edited.

kilemensi commented on src/lib/jsonSource/jsonSourceClient.js:
```diff
@@ -0,0 +1,60 @@
+            return [];
+          }
+        case "GET_KEY_PROMISES": {
+          const keyPromises = allPromises.filter((p) => p.keyPromise) ?? [];
+          return keyPromises.slice(0, limit);
```
Can chain these...
```suggestion
          const keyPromises = allPromises.filter((p) => p.keyPromise).slice(0, limit);
          return keyPromises;
```

filter will return empty array if no match was found I believe.

kilemensi reviewed (commented):
👍🏾

kilemensi reviewed (approved):
LGTM
